File: android/app/src/main/java/com/curiousbeagle/android/heartchart/ui/ErrorBanner.kt
package com.curiousbeagle.android.heartchart.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.curiousbeagle.android.heartchart.R
import com.curiousbeagle.android.heartchart.bluetooth.HeartRateMonitor
import com.curiousbeagle.android.heartchart.ui.theme.HeartChartTheme

/** Maps sensor errors to the localized copy users see. */
@Composable
fun errorMessage(error: HeartRateMonitor.SensorError): String = stringResource(
    when (error) {
        HeartRateMonitor.SensorError.CONNECTION_FAILED -> R.string.error_connection_failed
        HeartRateMonitor.SensorError.CONNECTION_LOST -> R.string.error_connection_lost
    }
)

/**
 * The single error surface: a red banner with the friendly message and a
 * Retry button that re-attempts the connection. Persists until the sensor
 * reconnects — the same pattern as TheMovieDBSwift.
 */
@Composable
fun ErrorBanner(error: HeartRateMonitor.SensorError, onRetry: () -> Unit) {
    Surface(
        color = MaterialTheme.colorScheme.error,
        shape = MaterialTheme.shapes.medium,
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp, vertical = 8.dp),
    ) {
        Row(
            modifier = Modifier.padding(start = 16.dp, end = 4.dp, top = 2.dp, bottom = 2.dp),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Icon(
                imageVector = Icons.Filled.Warning,
                contentDescription = null,
                tint = MaterialTheme.colorScheme.onError,
            )
            Text(
                text = errorMessage(error),
                style = MaterialTheme.typography.labelLarge,
                color = MaterialTheme.colorScheme.onError,
                modifier = Modifier.weight(1f),
            )
            TextButton(onClick = onRetry) {
                Text(
                    text = stringResource(R.string.retry),
                    color = MaterialTheme.colorScheme.onError,
                    style = MaterialTheme.typography.labelLarge,
                )
            }
        }
    }
}

@Preview(showBackground = true)
@Composable
private fun ErrorBannerPreview() {
    HeartChartTheme {
        ErrorBanner(HeartRateMonitor.SensorError.CONNECTION_LOST) {}
    }
}

File: iOS/HeartChart/HeartChart/Views/ErrorBanner.swift
//
//  ErrorBanner.swift
//  HeartChart
//

import SwiftUI

/// Maps sensor errors to the localized copy users see — the same pattern
/// as TheMovieDBSwift's error banner.
extension HeartRateMonitor.SensorError {
    var message: String {
        switch self {
        case .connectionFailed:
            String(localized: "error_connection_failed", defaultValue: "Couldn't connect to the sensor.")
        case .connectionLost:
            String(localized: "error_connection_lost", defaultValue: "Lost contact with the sensor.")
        }
    }
}

/// The single error surface: a red banner with the friendly message and a
/// Retry button that re-attempts the connection. Persists until the sensor
/// reconnects.
struct ErrorBanner: View {
    let error: HeartRateMonitor.SensorError
    let retry: () -> Void

    var body: some View {
        HStack(spacing: 12) {
            Label(error.message, systemImage: "exclamationmark.triangle.fill")
                .font(.subheadline.weight(.semibold))
                .multilineTextAlignment(.leading)
                .frame(maxWidth: .infinity, alignment: .leading)
            Button(String(localized: "retry", defaultValue: "Retry"), action: retry)
                .font(.subheadline.weight(.bold))
                .buttonStyle(.plain)
        }
        .foregroundStyle(.white)
        .padding(.vertical, 10)
        .padding(.horizontal, 16)
        .background(.red, in: RoundedRectangle(cornerRadius: 12))
        .padding(.horizontal)
        .accessibilityElement(children: .contain)
    }
}

#if DEBUG
#Preview("Lost contact") {
    ErrorBanner(error: .connectionLost) {}
}

#Preview("Connect failed") {
    ErrorBanner(error: .connectionFailed) {}
}
#endif
